integration/networking/port_mapping_linux_test.go

```diff
@@ -737,3 +739,122 @@ func TestDirectRoutingOpenPorts(t *testing.T) {
+
+	c := d.NewClientT(t)
+	defer c.Close()
+
+	testAccessDenied := func(t *testing.T, hostAddr string, nwOpts ...func(*networktypes.CreateOptions)) {
+		testutil.StartSpan(ctx, t)
+
+		const bridgeName = "nat-remote"
+		network.CreateNoError(ctx, t, c, bridgeName, append(nwOpts,
+			network.WithDriver("bridge"),
+			network.WithOption(bridge.BridgeName, bridgeName),
+		)...)
+		defer network.RemoveNoError(ctx, t, c, bridgeName)
+
+		// Set up a route on the attacker host to reach the docker bridge network
+		// created above, via the docker host.
+		if net.ParseIP(hostAddr).To4() != nil {
+			l3.Hosts["attacker"].Run(t, "ip", "route", "add", hostAddr+"/32", "via", "192.168.123.2", "dev", "eth0")
+			defer l3.Hosts["attacker"].Run(t, "ip", "route", "delete", hostAddr+"/32", "via", "192.168.123.2", "dev", "eth0")
+		} else {
+			l3.Hosts["attacker"].Run(t, "ip", "route", "add", hostAddr+"/128", "via", "fde8:19ff:6e09::2", "dev", "eth0")
+			defer l3.Hosts["attacker"].Run(t, "ip", "route", "delete", hostAddr+"/128", "via", "fde8:19ff:6e09::2", "dev", "eth0")
+		}
+
+		const hostPort = "5000"
+
+		// Create the victim container, with a non-empty / non-unspecified
+		// HostIP in its port binding.
+		serverID := container.Run(ctx, t, c,
+			container.WithName(sanitizeCtrName(t.Name()+"-server")),
+			container.WithCmd("nc", "-lup", "5000"),
+			container.WithExposedPorts("5000/udp"),
+			container.WithPortMap(nat.PortMap{"5000/udp": {{HostIP: hostAddr, HostPort: hostPort}}}),
+			container.WithNetworkMode(bridgeName))
+		defer c.ContainerRemove(ctx, serverID, containertypes.RemoveOptions{Force: true})
+
+		// Send a UDP datagram to the published port, from the attacker host
+		// running in a separate L2 segment.
+		//
+		// Here UDP is preferred, because it's a one-way, connectionless
+		// protocol. With TCP the three-way handshake has to be completed
+		// before sending a payload. But since some of the test below will try
+		// to spoof the loopback address, the 'attacker host' will drop the
+		// SYN-ACK by default (because the source addr will be considered
+		// invalid / non-routable). This would require further tuning to make
+		// it work. But with UDP, this problem doesn't exist - the payload can
+		// be sent straight away.
+		l3.Hosts["attacker"].Do(t, func() {
+			t.Logf("Sending a datagram to %s:%s", hostAddr, hostPort)
+
+			// Send a request to the victim container from the attacker host.
+			icmd.RunCommand("/bin/sh", "-c", fmt.Sprintf("echo foobar | nc -w1 -u %s %s", hostAddr, hostPort)).Assert(t, icmd.Success)
+		})
+
+		// Assert on outputs
+		logReader, err := c.ContainerLogs(ctx, serverID, containertypes.LogsOptions{ShowStdout: true})
+		assert.NilError(t, err)
+		defer logReader.Close()
+
+		var actualStdout bytes.Buffer
+		_, err = stdcopy.StdCopy(&actualStdout, nil, logReader)
+		assert.NilError(t, err)
+
+		stdOut := strings.TrimSpace(actualStdout.String())
+		assert.Assert(t, !strings.Contains(stdOut, "foobar"))
+	}
+
```

Comment:
It'd probably be worth adding a test that's expected to succeed? (By sending to the "eth-test" addresses I think?)

That'd help to confirm the packet's not getting through in the other tests because of the new rule - rather than the container log read being too quick, or some other test problem?

Reply:
Maybe also run the tests twice, with the escape hatch env-var set and unset?

Reply:
The tests now run with and without the escape hatch, and all test cases also run the 'happy case' to make sure the test is working as expected.